src/bin/rekordcrate-pdb.rs

```diff
@@ -6,34 +6,48 @@
 //
 // SPDX-License-Identifier: MPL-2.0
 
-use rekordcrate::pdb::{Header, RowGroup};
+use binrw::{BinRead, ReadOptions};
+use rekordcrate::pdb::{Header, Page, Row, RowOffset};
+use std::io::{Seek, SeekFrom};
 
 fn main() {
     let path = std::env::args().nth(1).expect("no path given");
-    let data = std::fs::read(&path).expect("failed to read file");
-    let (_, header) = Header::parse(&data).expect("failed to parse header");
+    let mut reader = std::fs::File::open(&path).expect("failed to open file");
+    let header = Header::read(&mut reader).expect("failed to parse pdb file");
 
     println!("{:#?}", header);
 
     for (i, table) in header.tables.iter().enumerate() {
         println!("Table {}: {:?}", i, table.page_type);
-        for page_index in table.page_indices(&header, &data) {
-            let (_, page) = header
-                .page(&data, &page_index)
-                .expect("failed to parse page");
+        for page in header
+            .read_pages(
+                &mut reader,
+                &ReadOptions::default(),
+                (&table.first_page, &table.last_page),
+            )
+            .unwrap()
+            .into_iter()
+        {
+            let page_offset = header.page_offset(&page.page_index);
             println!("  {:?}", page);
-            assert_eq!(page.page_index, page_index);
-            let page_offset = header.page_offset(&page_index).unwrap();
-            let page_data = &data[page_offset..];
-            page.row_groups(page_data, header.page_size)
-                .for_each(|row_group| {
-                    println!("    {:?}", row_group);
-                    let RowGroup(row_offsets) = row_group;
-                    for row_offset in row_offsets {
-                        let (_, row) = page.row(page_data, &row_offset).unwrap();
-                        println!("      {:?}", row);
-                    }
-                })
+            page.row_groups.iter().for_each(|row_group| {
+                println!("    {:?}", row_group);
+                for &RowOffset(row_offset) in row_group.present_rows() {
+                    let abs_offset: u64 = page_offset
+                        + u64::try_from(Page::HEADER_SIZE).unwrap()
+                        + u64::from(row_offset);
+                    reader
+                        .seek(SeekFrom::Start(abs_offset))
+                        .expect("failed to seek to row offset");
+                    let row = Row::read_options(
+                        &mut reader,
+                        &ReadOptions::default(),
+                        (page.page_type.clone(),),
+                    )
+                    .expect("failed to parse row");
+                    println!("      {:?}", row);
```

Comment:
having to manually mess with the offset is highly unsafe and not something I would make part of the API in any shape. First of all because a lot could go wrong when the consumer is manually in charge of juggling all these types, and the resulting boilerplate required to make this work is also less than ideal. Basically, its easy to misuse and hard to use correctly. https://www.oreilly.com/library/view/97-things-every/9780596809515/ch55.html

Reply:
Also, making binrw part of the public API is not good either.

Reply:
> having to manually mess with the offset is highly unsafe and not something I would make part of the API in any shape. First of all because a lot could go wrong when the consumer is manually in charge of juggling all these types, and the resulting boilerplate required to make this work is also less than ideal. Basically, its easy to misuse and hard to use correctly. https://www.oreilly.com/library/view/97-things-every/9780596809515/ch55.html

I agree and already have some code that gets rid of this. The downside is that it just parse all rows in a page at once, not lazily.

> Also, making binrw part of the public API is not good either.

There is only alternative I can think of (other than to not use binrw at all) would be to copy all parsed data into another struct that does not implement `BinRead`/`BinWrite`. If the `trait` and the `struct` is `pub`, so is the `impl`. We can document that using `binrw` is an implementation detail and depending on it is not recommended though.

Anyway, I wouldn't worry too much about the public API yet. We are still in the experimentation phase, and hiding binrw stuff just makes the code complicated without much gain.  Later on, we need to think how to make serialization possible without in-depth knowledge of the format without creating struct instances by hand anyway. Whoever uses this library probably doesn't care about the various string types, page size and how many rows there are in a row group.

Reply:
Don't get me wrong, I agree that completely eliminating binrw from our public API is not worth the effort, but the amount of boilerplate and types needed for the basic act of just reading out all rows is not good. If we only require the API consumer to make use of BinRead/BinWrite, thats fine IMO. 

> I agree and already have some code that gets rid of this. The downside is that it just parse all rows in a page at once, not lazily. Didn't want to put it into this PR because it's already a pretty large diff.

Ok, then I won't dwell on this issue in this PR any longer.

Reply:
Here's the commit btw: https://github.com/Holzhaus/rekordcrate/commit/88c87c83f87ede77adb8fbe4fcb8ec46c06613b7